src/main/java/dev/openfeature/sdk/EventEmitter.java

```diff
@@ -0,0 +1,108 @@
+package dev.openfeature.sdk;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import dev.openfeature.sdk.internal.AutoCloseableLock;
+import dev.openfeature.sdk.internal.AutoCloseableReentrantReadWriteLock;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Event emitter construct to be used by providers.
+ */
+@Slf4j
+public class EventEmitter {
+
+    private static final ExecutorService taskExecutor = Executors.newCachedThreadPool();
+    private final Map<ProviderEvent, List<Consumer<EventDetails>>> handlerMap;
+    private AutoCloseableReentrantReadWriteLock handlersLock = new AutoCloseableReentrantReadWriteLock();
+
+    /**
+     * Construct a new EventEmitter.
+     */
+    public EventEmitter() {
+        handlerMap = new ConcurrentHashMap<ProviderEvent, List<Consumer<EventDetails>>>();
+        handlerMap.put(ProviderEvent.PROVIDER_READY, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_CONFIGURATION_CHANGED, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_ERROR, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_STALE, new ArrayList<>());
```

Comment:
We could construct this map lazily if we wanted.

Reply:
I think you're right, it's worth doing. I'll make things as lazy as possible to save resources. Until handlers are actually defined we can use almost no space.

Reply:
I've made the initialization of the store lazy, for each client, but not each event type, meaning that I make space for all event types for a client if a single one is added. I think the tiny memory saved here is not work the complexity.